### DISCUSSION TITLE
How did you come up with the delegate-arg pattern for method chaining?

### DISCUSSION THREAD

**Discussion by pha3z** [QuestPDF/QuestPDF#76] [General]
I began scrutinizing the delegate-arg pattern to understand what really makes it tick.

You have this example code on the readme:
```
container.Row(row =>
{
    row.RelativeColumn().Stack(stack =>
    {
        stack.Item().Text($"Invoice #{Model.InvoiceNumber}", titleStyle);

        stack.Item().Text(text =>
        {
            text.Span("Issue date: ", TextStyle.Default.SemiBold());
            text.Span($"{Model.IssueDate:d}");
        });

        stack.Item().Text(text =>
        {
            text.Span("Due date: ", TextStyle.Default.SemiBold());
            text.Span($"{Model.DueDate:d}");
        });
    });
    
    row.ConstantColumn(100).Height(50).Placeholder();
});
```

Without the delegate argments, the code would have to be rewritten as something like this:
```
var row = container.Row();
var stack = row.RelativeColumn().Stack();
stack.Item().Text($"Invoice #{Model.InvoiceNumber}", titleStyle);
stack.Item().Text()
    .Span("Issue date: ", TextStyle.Default.SemiBold());
    .Span($"{Model.IssueDate:d}");
stack.Item().Text()
    .Span("Due date: ", TextStyle.Default.SemiBold());
    .Span($"{Model.DueDate:d}");
row.ConstantColumn(100).Height(50).Placeholder();
```

Pitfalls of the non-delegate version: 

- We're forced to return objects to variables in order to keep working with them.
- The code gets harder to maintain because you only get chaining at the lowest depth, which makes it non-intuitive to chain at all since you have to mentally think about what depth you're at and whether it would be beneficial to chain there.
- Since we're creating variables, we have to think about the program procedurally (whether or not the variables get used later).  In the delegate-variant, you never even have to think about it.
- Indentation becomes non-trivial.  Have to make more challenging decisions about formatting. Yuck.

Delegate variant eliminates all the pitfalls above. I'm curious to know if I've interpretted correctly.  And also, where did you get the inspiration for this from?  I don't think I would have reached for this intuitively in any API I design, but now that I get it, it seems awesome!

**Comment by MarcinZiabek** [comment]
I am a huge enthusiast of the functional programming paradigm. I enjoy using higher-order functions to improve the code style and usability in C#. The first place where I consciously observed this particular pattern was the query builder in the ElasticSearch client. Later, at my job, I started building various libraries: to configure generalized frontend pages and handle events from them, to generate reports or handle incoming events from external resources. Making such configurations in pure JSON (or any other format) is not very safe. Therefore, I decided to build such a configuration with Fluent API (therefore with static code analysis, IntelliSense, etc.), functional programming and expression trees. As we have very complex, deeply nested hierarchies, I needed a way to reduce nesting and make such configuration code as concise and possible. Several iterations later, many trials and errors and we have API design in QuestPDF 😁

Please notice that QuestPDF follows a very strict rule: every component needs to be as simple as possible, composition over complexity.  This was inspired by Flutter. Therefore, instead of creating a complex element that handles border, padding, alignment, etc. (like in HTML), I have created several elements responsible for one functionality at a time. This makes the implementation manageable but increases the number of required elements substantially. Without a proper way of chaining and reducing nesting, working with such code would be difficult. 

Personally, I would like to see a UI library that follows a similar pattern 😝 Scrolling through all properties/events/methods in WPF is a nightmare...

**Comment by pha3z** [comment]
Fascinating background on how you arrived at the design.  Thank you for sharing!

And yes..... Wpf (and frankly every one of the XAML frameworks) are a total disaster.  1 month of trying WPF made me denounce the entire ecosystem.  I go elsewhere for GUI work.

**Reply by MarcinZiabek** [reply]
Personally, I really enjoyed working with WPF when creating my bachelors and masters theses. It allows creating really complex and advanced software with high quality. However, the complexity of each layout element in WPF (and many other UI libraries) makes me believe that composition is really more important than inheritance. When a class has hundreds of properties, it is not good, is it? 

Instead of teaching OOP, we should consider showing functional algebra. Not saying OOP is wrong but it certainly has many caveats that beginners are not ready for.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
